bittensor/utils/subtensor.py

```diff
@@ -0,0 +1,137 @@
+# The MIT License (MIT)
+# Copyright © 2021 Yuma Rao
+# Copyright © 2022 Opentensor Foundation
+# Copyright © 2023 Opentensor Technologies Inc
+
+# Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
+# documentation files (the “Software”), to deal in the Software without restriction, including without limitation
+# the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software,
+# and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
+
+# The above copyright notice and this permission notice shall be included in all copies or substantial portions of
+# the Software.
+
+# THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO
+# THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+# THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+# OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+# DEALINGS IN THE SOFTWARE.
+
+"""Module providing common helper functions for working with Subtensor."""
+
+import json
+import logging
+import os
+from typing import Dict, Optional, Union, Any
+
+from substrateinterface.base import SubstrateInterface
+
+_logger = logging.getLogger("subtensor.errors_handler")
+
+_USER_HOME_DIR = os.path.expanduser("~")
+_BT_DIR = os.path.join(_USER_HOME_DIR, ".bittensor")
+_ERRORS_FILE_PATH = os.path.join(_BT_DIR, "subtensor_errors_map.json")
+_ST_BUILD_ID = "subtensor_build_id"
+
+# Create directory if it doesn't exist
+os.makedirs(_BT_DIR, exist_ok=True)
+
+
+# Pallet's typing class `PalletMetadataV14` is defined only at
+# https://github.com/polkascan/py-scale-codec/blob/master/scalecodec/type_registry/core.json#L1024
+# A class object is created dynamically at runtime.
+# Circleci linter complains about string represented classes like 'PalletMetadataV14'.
+def _get_errors_from_pallet(pallet) -> Optional[Dict[str, Dict[str, str]]]:
+    """Extracts and returns error information from the given pallet metadata.
+
+    Args:
+        pallet (PalletMetadataV14): The pallet metadata containing error definitions.
+
+    Returns:
+        dict[str, str]: A dictionary of errors indexed by their IDs.
+
+    Raises:
+        ValueError: If the pallet does not contain error definitions or the list is empty.
+    """
+    if not hasattr(pallet, "errors") or not pallet.errors:
+        _logger.warning(
```

Comment:
Should we use bittensor's logger here instead of a new logger?

Reply:
Good question!
If I use `from bittensor import logging` I get the following error. Since we were discussing offline, let's take these changes forward and replace the imports during the refactoring.
```python
test_subtensor.py:22: in <module>
    import bittensor.utils.subtensor as st_utils
../../../bittensor/__init__.py:289: in <module>
    from .subtensor import subtensor as subtensor
../../../bittensor/subtensor.py:88: in <module>
    from .utils.subtensor import get_subtensor_errors
../../../bittensor/utils/subtensor.py:25: in <module>
    from bittensor import logging as _logger
E   ImportError: cannot import name 'logging' from partially initialized module 'bittensor' (most likely due to a circular import) (/Users/roman/Python/bittensor/bittensor/__init__.py)
```